modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java

```diff
@@ -197,13 +202,100 @@ public T next() {
         };
     }
 
+    /**
+     * Create a lazy concatenation of iterators.
+     *
+     * <p>NOTE: {@link Iterator#remove} - not supported.
+     *
+     * @param iterators Iterators.
+     * @param <T> Type of the elements.
+     * @return Concatenation of iterators.
+     */
+    @SafeVarargs
+    public static <T> Iterator<T> concat(@Nullable Iterator<? extends T>... iterators) {
+        if (iterators == null || iterators.length == 0) {
+            return emptyIterator();
+        }
+
+        return new Iterator<>() {
+            /** Current index at {@code iterators}. */
+            int idx = 0;
+
+            /** Current iterator. */
+            Iterator<? extends T> curr = emptyIterator();
+
+            /** {@inheritDoc} */
+            @Override
+            public boolean hasNext() {
+                while (!curr.hasNext() && idx < iterators.length) {
+                    curr = iterators[idx++];
+                }
+
+                return curr.hasNext();
+            }
+
+            /** {@inheritDoc} */
+            @Override
+            public T next() {
+                if (!hasNext()) {
+                    throw new NoSuchElementException();
+                } else {
+                    return curr.next();
+                }
+            }
+        };
+    }
+
+    /**
+     * Create a lazy concatenation of iterators.
+     *
+     * <p>NOTE: {@link Iterator#remove} - not supported.
+     *
+     * @param iterators Iterators.
+     * @param <T> Type of the elements.
+     * @return Concatenation of iterators.
+     */
+    public static <T> Iterator<T> concat(@Nullable Collection<Iterator<? extends T>> iterators) {
```

Comment:
Same question, this is literally a `flatMap(Function.id())` operation

Reply:
I wrote above about the impossibility of using streams now.